### ISSUE TITLE
The options parameter is sticky across calls to QnAMaker.GetAnswersAsync.

### ISSUE THREAD

**Issue by negativeeddy** [microsoft/botbuilder-dotnet#2018]
## Version
4.4.4

## Describe the bug
The documentation for the *options* parameter on QnAMaker.GetAnswersAsync is incorrect. It states

> The options for the QnA Maker knowledge base. If null, constructor option is used for this instance

When in fact it will use the most recently passed in set of options as well. The options parameter is sticky across calls to GetAnswersAsync.

## To Reproduce
1. Make a call to QnAMaker.GetAnswersAsync with the options set with a metadata filter.
2. Then make a call to QnAMaker.GetAnswersAsync with options set to null

Both of these calls return the same query because the options is persisted into the second call.

## Expected behavior
The first call should have the filtered results and the second should have the normal result

## Additional context
I'm not sure if this is a bug or a documentation error, but the expected behavior does not match the documented behavior.

Additionally this affects the Virtual Assistant which stores all the QnA services as singletons. So metadata filters leak impact the next bot request as well.
[bug]

**Comment by tdurnford** [comment]
Seems to be slightly related to this issue https://github.com/microsoft/botframework-sdk/issues/5405

**Comment by Zerryth** [comment]
@negativeeddy could you please provide examples of how you're getting this behavior? I'm not able to repro what you're describing, especially since this should've been fixed in 4.4 release.
___
### Here's what I've done:
_Note that although this is inside a unit test for qnamaker, I'm actually calling a "real" qna api endpoint instead of a mock_

#### KB
![image](https://user-images.githubusercontent.com/35248895/59524323-66a07900-8e88-11e9-9841-c0c597368bf0.png)

#### QnA Maker Setup

* qna constructor has options (Top 10)
* 1st call to `GetAnswersAsync()` has a different set of options to query KB with (StrictFilters, only Top 1 result)
* 2nd call to `GetAnswersAsync()` has null options

```cs
            var constructorOptions = new QnAMakerOptions
            {
                Top = 10,
            };

            var qna = new QnAMaker(hardcodedEndpoint, constructorOptions);

            var queryOptions = new QnAMakerOptions
            {
                StrictFilters = new Metadata[]
                {
                    new Metadata() { Name = "movie", Value = "disney" },
                },
                Top = 1,
            };
            var resultsWithQueryOptions = await qna.GetAnswersAsync(GetContext("up"), queryOptions);
            var resultsNullOptions = await qna.GetAnswersAsync(GetContext("up"), null);
```
#### Results
![image](https://user-images.githubusercontent.com/35248895/59524716-78cee700-8e89-11e9-9904-3763008d8236.png)

* results with queryOptions returns only 1 result, as specified by the options passed into `GetAnswersAsync()`, overwriting the constructor options that specified to return top 10 results
* results with null options returns 4 qna results, as it is no longer constrained by "Top 1" option, but rather takes "Top 10" as specified in constructor. Also included in the 4 results is the answer for "up" that does _not_ include any metadata, so we see that StrictFilters are also not a part of options as well, just like in constructor options

**Comment by Zerryth** [comment]
Closing, as of no response after a week. Feel free to comment on how exactly this issue was repro'd and we can always reopen

**Comment by negativeeddy** [comment]
sorry for the delay in response. Unfortunately my original description was not accurate after going back and reviewing it. This code demonstrates the problem. If the original construction options is set with a metadata filter, you can not remove the Metadata filter when calling GetAnswersAsync. My original testing, at some point, had the filter set in both the construction of the service and the first use, so I thought it was the first use that was the problem.

```C#
var options = new QnAMakerOptions
{
    MetadataBoost = new Metadata[]
    {
                    new Metadata() { Name = "artist", Value = "drake" },
    },
    Top = 1,
};

var qna = GetQnAMaker(
    interceptHttp,
    new QnAMakerEndpoint
    {
        KnowledgeBaseId = _knowlegeBaseId,
        EndpointKey = _endpointKey,
        Host = _hostname,
    },
    options);

var results = await qna.GetAnswersAsync(GetContext("who loves me?"));

var options2 = new QnAMakerOptions
{
    MetadataBoost = new Metadata[0],
    Top = 1,
};

results = await qna.GetAnswersAsync(GetContext("who loves me?"), options2);

var options3 = new QnAMakerOptions
{
    MetadataBoost = null,
    Top = 1,
};

results = await qna.GetAnswersAsync(GetContext("who loves me?"), options3);

```
if you replace the QnaMaker_ReturnsAnswerWithMetadataBoost test project code with this code and step into QnaMaker.HydrateOptions() you will see in the second pass it ignores the empty array of metadata. In all three calls HydrateOptions() ends up with the original metadata. So there is no way to replace the default metadata option with an empty set. You can only replace the metadata with a different set of metadata. I would expect the second call to replace it with the no metadata value. 

FYI - my scenario is that I have a KB which has many questions. We want to add filters (based on user role) to some of them. In order to not have to tag all the questions, we call it once with the metadata and if no answer is matched we call again with an empty filter. I've worked around this for now by changing the QnAService in the bot from singleton to transient.

**Comment by Zerryth** [comment]
> there is no way to replace the default metadata option with an empty set

Ahh, I see. This is true. Thanks for providing examples - reopened the issue

**Comment by Zerryth** [comment]
Ok, spoke with the SDK team, and we cannot change the behavior of how options currently function--that is, with the constructor's options as the base options to populate queries with, and with query options being able to be passed in via `GetAnswersAsync()` to replace/add options different from what's in the constructor. The behavioral change would be breaking

Expected Behavior:
So say somebody wanted all their queries to return Top 10 responses, but for their with GetAnswers, they also wanted to specifically wanted to use `StrictFilters` -- the options will work to both apply the `StrictFilters` and return the top 10 results within threshold. If they wanted to change it to Top 1 results in the GetAnswers call, then that is possible to do so, and the 1 would overwrite the 10.
It is only in situations where you place 0 or null where it will default to constructor options over GetAnswers options.
-- To get down into the nitty gritty, most of the time, one defines only the options that they want to apply, so if we overwrote options passed into GetAnswers completely, it would overwrite the "base" options that were passed into the constructor (like Top in our example above, if we only specified StrictFilters)
___
### How to Write Your Calls with MetadataBoost Options Properly
So what would work for you, and allow you to keep qna as a singleton would be to actually just put your **`MetadataBoost` option into your calls with `GetAnswers` and leave the `MetadataBoost` options in the constructor null**.

Since you're making 2 calls with `GetAnswers` anyways, it would look like this

* Call 1: `GetAnswers` with `MetadataBoost` specified in the query
* Call 2 IF Call 1 received no results -- call `GetAnswers` with no `MetadataBoost` specified in options. Null options default to what's in the constructor--what's in the constructor is null `MetadataBoost`, so therefore it will not have `MetadataBoost` option passed into your 2nd `GetAnswers` call
___
### QnA API Intended Use Note

One last thing regarding `MetadataBoost` specifically.

You mentioned that you chose the route of not tagging all the Q&A pairs with metadata, simply because of the sheer size of the KB, and thus use this `MetadataBoost` scenario you described.

Ultimately the best option with using QnA Maker now is actually to go the route of actually tagging your Q&A pairs with metadata in your KB, and querying with `StrictFilters`, **as `MetadataBoost` is not included in the GA QnA Maker v4.0 API**.

In [this post regarding "Metadata boost results"](https://github.com/Microsoft/BotBuilder-CognitiveServices/issues/132#issuecomment-488597213), @pchoudhari, who works on the QnA team directly, actually responded that it was not on the team's roadmap to bring the MetadataBoost feature back any time soon, and even suggests the StrictFilters approach as well.
